### DISCUSSION TITLE
How to access 'hidden' fields during migration

### DISCUSSION THREAD

**Discussion by JanezStupar** [simolus3/drift#2725] [General]
I am working on a [CRDT plugin for Drift](https://pub.dev/packages/drift_crdt). 

One issue I ran into is the matter of migrations.

Long story short is, tables that are meant to be replicated have 4 additional columns added (`is_deleted`, `hlc`, `node_id` and `modified`).

Those 4 columns are added to the table under the hood by the plugin. They are also obscured from the user for UX reasons.

However when tables get migrated it is very important that those 4 fields get their values copied over too.

Is there a way to let migration and only migration know these fields exist?

**Comment by simolus3** [comment]
> However when tables get migrated it is very important that those 4 fields get their values copied over too.

I don't think this is possible with drift APIs alone unless you create your own `GeneratedColumn` instances reflecting these internal columns.

These columns are hidden in a pretty low level by using the `sqlparser` package to intercept queries, so even a `SELECT *` wouldn't find them.

**Reply by JanezStupar** [reply]
So if I understand correctly, I should manually upgrade the generated migration script to add the hidden columns via the GeneratedColumn instances, right?

I am not completely done yet, but this is what I have started doing and it looks promising.

I am asking because you might have had something different in mind that I missed.

**Reply by simolus3** [reply]
Right - you don't need to add them to any `CREATE TABLE` statements since the SQL interceptor is doing that at a lower level, but you need them when writing `insert from select` statements to copy data over during a table migration.

**Reply by JanezStupar** [reply]
I got the migrations to work and documented it for future reference.

https://pub.dev/packages/drift_crdt#drift-migrations
